// @logicflow/extension/lib/bpmn/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceFlowModel = exports.SequenceFlowView = exports.ServiceTaskModel = exports.ServiceTaskView = exports.UserTaskModel = exports.UserTaskView = exports.ExclusiveGatewayModel = exports.ExclusiveGatewayView = exports.EndEventModel = exports.EndEventView = exports.StartEventView = exports.StartEventModel = exports.BpmnElement = void 0;
var constant_1 = require("./constant");
var events_1 = require("./events");
Object.defineProperty(exports, "StartEventView", { enumerable: true, get: function () { return events_1.StartEventView; } });
Object.defineProperty(exports, "StartEventModel", { enumerable: true, get: function () { return events_1.StartEventModel; } });
Object.defineProperty(exports, "EndEventView", { enumerable: true, get: function () { return events_1.EndEventView; } });
Object.defineProperty(exports, "EndEventModel", { enumerable: true, get: function () { return events_1.EndEventModel; } });
var flow_1 = require("./flow");
Object.defineProperty(exports, "SequenceFlowView", { enumerable: true, get: function () { return flow_1.SequenceFlowView; } });
Object.defineProperty(exports, "SequenceFlowModel", { enumerable: true, get: function () { return flow_1.SequenceFlowModel; } });
var gateways_1 = require("./gateways");
Object.defineProperty(exports, "ExclusiveGatewayView", { enumerable: true, get: function () { return gateways_1.ExclusiveGatewayView; } });
Object.defineProperty(exports, "ExclusiveGatewayModel", { enumerable: true, get: function () { return gateways_1.ExclusiveGatewayModel; } });
var tasks_1 = require("./tasks");
Object.defineProperty(exports, "UserTaskView", { enumerable: true, get: function () { return tasks_1.UserTaskView; } });
Object.defineProperty(exports, "UserTaskModel", { enumerable: true, get: function () { return tasks_1.UserTaskModel; } });
Object.defineProperty(exports, "ServiceTaskView", { enumerable: true, get: function () { return tasks_1.ServiceTaskView; } });
Object.defineProperty(exports, "ServiceTaskModel", { enumerable: true, get: function () { return tasks_1.ServiceTaskModel; } });
var BpmnElement = {
    pluginName: 'bpmnElement',
    install: function (lf) {
        lf.setTheme(constant_1.theme);
        lf.register(events_1.StartEvent);
        lf.register(events_1.EndEvent);
        lf.register(gateways_1.ExclusiveGateway);
        lf.register(tasks_1.UserTask);
        lf.register(tasks_1.ServiceTask);
        // 支持自定义bpmn元素的边
        if (!lf.options.customBpmnEdge) {
            lf.register(flow_1.SequenceFlow);
            lf.setDefaultEdgeType('bpmn:sequenceFlow');
        }
    },
};
exports.BpmnElement = BpmnElement;
