// @logicflow/extension/lib/mindmap/fakerRoot.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@logicflow/core");
var MarkRootModel = /** @class */ (function (_super) {
    __extends(MarkRootModel, _super);
    function MarkRootModel() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MarkRootModel.extendKey = 'MarkRootModel';
    return MarkRootModel;
}(core_1.RectNodeModel));
var MarkRootView = /** @class */ (function (_super) {
    __extends(MarkRootView, _super);
    function MarkRootView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MarkRootView.prototype.getShape = function () {
        return null;
    };
    return MarkRootView;
}(core_1.BaseNode));
var MarkRoot = {
    type: 'faker:root',
    view: MarkRootView,
    model: MarkRootModel,
};
exports.default = MarkRoot;
