// @logicflow/extension/es/bpmn/flow/SequenceFlow.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
import { PolylineEdge, PolylineEdgeModel } from '@logicflow/core';
import { getBpmnId } from '../getBpmnId';
var SequenceFlowModel = /** @class */ (function (_super) {
    __extends(SequenceFlowModel, _super);
    function SequenceFlowModel(data, graphModel) {
        if (!data.id) {
            data.id = "Flow_".concat(getBpmnId());
        }
        return _super.call(this, data, graphModel) || this;
    }
    SequenceFlowModel.extendKey = 'SequenceFlowModel';
    return SequenceFlowModel;
}(PolylineEdgeModel));
export { SequenceFlowModel };
var SequenceFlowView = /** @class */ (function (_super) {
    __extends(SequenceFlowView, _super);
    function SequenceFlowView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SequenceFlowView.extendKey = 'SequenceFlowEdge';
    return SequenceFlowView;
}(PolylineEdge));
export { SequenceFlowView };
export var SequenceFlow = {
    type: 'bpmn:sequenceFlow',
    view: SequenceFlowView,
    model: SequenceFlowModel,
};
export default SequenceFlow;
