// @logicflow/extension/lib/bpmn/flow/SequenceFlow.js
"use strict";
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (Object.prototype.hasOwnProperty.call(b, p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        if (typeof b !== "function" && b !== null)
            throw new TypeError("Class extends value " + String(b) + " is not a constructor or null");
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
Object.defineProperty(exports, "__esModule", { value: true });
exports.SequenceFlow = exports.SequenceFlowView = exports.SequenceFlowModel = void 0;
var core_1 = require("@logicflow/core");
var getBpmnId_1 = require("../getBpmnId");
var SequenceFlowModel = /** @class */ (function (_super) {
    __extends(SequenceFlowModel, _super);
    function SequenceFlowModel(data, graphModel) {
        if (!data.id) {
            data.id = "Flow_".concat((0, getBpmnId_1.getBpmnId)());
        }
        return _super.call(this, data, graphModel) || this;
    }
    SequenceFlowModel.extendKey = 'SequenceFlowModel';
    return SequenceFlowModel;
}(core_1.PolylineEdgeModel));
exports.SequenceFlowModel = SequenceFlowModel;
var SequenceFlowView = /** @class */ (function (_super) {
    __extends(SequenceFlowView, _super);
    function SequenceFlowView() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    SequenceFlowView.extendKey = 'SequenceFlowEdge';
    return SequenceFlowView;
}(core_1.PolylineEdge));
exports.SequenceFlowView = SequenceFlowView;
exports.SequenceFlow = {
    type: 'bpmn:sequenceFlow',
    view: SequenceFlowView,
    model: SequenceFlowModel,
};
exports.default = exports.SequenceFlow;
